rdr_service/tools/tool_libs/genomic_utils.py

```diff
@@ -673,6 +674,172 @@ def update_genomic_set_member_state(self, member, state):
         genomic_set_member_update(member.id)
 
 
+class ChangeCollectionTube(GenomicManifestBase):
+    """Class to update a genomic_set_member with a different collection tube id."""
+
+    def __init__(self, args, gcp_env: GCPEnvConfigObject):
+        super(ChangeCollectionTube, self).__init__(args, gcp_env)
+
+    def run(self):
+        """
+        Main program process
+        :return: Exit code value
+        """
+
+        _logger.info("Running Genomic Collection Tube tool")
+        if self.args.dryrun:
+            _logger.info("Running in dryrun mode. No actual updates to data.")
+            self.msg = "Would update"
+
+        # Validate Aruguments
+        if not os.path.exists(self.args.file):
+            _logger.error(f'File {self.args.file} was not found.')
+            return 1
+
+        # Activate the SQL Proxy
+        self.gcp_env.activate_sql_proxy()
+        self.dao = GenomicSetMemberDao()
+
+        # Document these changes in the dev_note field
+        dev_note = input("Please enter a developer note for these updates:\n")
+        if dev_note in (None, ""):
+            _logger.error('A developer note is required for these changes.')
+            return 1
+
+        # Update gsm IDs from file
+        with open(self.args.file, encoding='utf-8-sig') as f:
+            csvreader = csv.reader(f, delimiter=",")
+
+            # Groups for output/logging
+            nonexistent_bids = []
+            invalid_tubes = []
+            already_existing_tubes_same_member = []
+            already_existing_tubes_diff_member = []
+
+            # Iterate through each bid-tube_id pair
+            for line in csvreader:
+                _bid = line[0]
+                _new_tube_id = line[1]
+
+                # Check that supplied new_tube_id is valid and associated to bid
+                valid_tube = self._validate_tube_id(_bid, _new_tube_id)
+                if not valid_tube:
+                    invalid_tubes.append((_bid, _new_tube_id))
+
+                # lookup member for each genome type
+                for _genome_type in (config.GENOME_TYPE_WGS, config.GENOME_TYPE_ARRAY):
+                    member = self.dao.get_member_from_biobank_id(_bid, _genome_type)
+
+                    if member is None:
+                        # Member with BID doesn't exist, skip it and log it.
+                        nonexistent_bids.append((_bid, _genome_type))
+                        continue
+
+                    else:
+                        member.devNote = dev_note
```

Comment:
I could be wrong, but I think at this point the session that loaded `member` is closed so setting devNote isn't guaranteed to be saved. Later in `update_genomic_set_member_collection_tube` it gets merged to a new session, so I think that path would save the devNote. But if the else-block is followed then I think the devNote would not be saved.

This probably isn't a big deal, and it might be doing what you want it to anyway.

Reply:
At that point, the member object should be in memory with the `devNote` attribute set. I'm passing a reference to the member object to the `update_genomic_set_member_collection_tube` method. I haven't run into any issues testing this field on sandbox. 
However, I think I need to move it into the next condition: `if new_tube_member is None and valid_tube:`
Otherwise, when it compares `new_tube_member` to `member` on line 749, the devNote attribute may cause this condition to be false.

Reply:
👍